db/db_impl/db_impl_follower.cc

```diff
@@ -0,0 +1,271 @@
+          1000);
+      ;
+    }
+    if (!s.ok()) {
+      ROCKS_LOG_INFO(immutable_db_options_.info_log, "Catch up unsuccessful");
+    }
+  }
+}
+
+Status DBImplFollower::CloseImpl() {
+  if (catch_up_thread_) {
+    stop_requested_.store(true);
+    catch_up_thread_->join();
+  }
+
+  return DBImpl::Close();
+}
+
+Status DB::OpenAsFollower(const Options& options, const std::string& dbname,
+                          const std::string& leader_path, DB** dbptr) {
+  *dbptr = nullptr;
+
+  DBOptions db_options(options);
+  ColumnFamilyOptions cf_options(options);
+  std::vector<ColumnFamilyDescriptor> column_families;
+  column_families.emplace_back(kDefaultColumnFamilyName, cf_options);
+  std::vector<ColumnFamilyHandle*> handles;
+
+  Status s = DB::OpenAsFollower(db_options, dbname, leader_path,
+                                column_families, &handles, dbptr);
+  if (s.ok()) {
+    assert(handles.size() == 1);
+    delete handles[0];
+  }
+  return s;
+}
+
+Status DB::OpenAsFollower(
+    const DBOptions& db_options, const std::string& dbname,
+    const std::string& src_path,
+    const std::vector<ColumnFamilyDescriptor>& column_families,
+    std::vector<ColumnFamilyHandle*>* handles, DB** dbptr) {
+  *dbptr = nullptr;
+
+  FileSystem* fs = db_options.env->GetFileSystem().get();
+  IOStatus io_s = fs->CreateDirIfMissing(dbname, IOOptions(), nullptr);
+  if (!io_s.ok()) {
+    return static_cast<Status>(io_s);
+  }
+  std::unique_ptr<Env> new_env(new CompositeEnvWrapper(
+      db_options.env, NewOnDemandFileSystem(db_options.env->GetFileSystem(),
+                                            src_path, dbname)));
+
+  DBOptions tmp_opts(db_options);
+  Status s;
+  tmp_opts.env = new_env.get();
+  if (nullptr == tmp_opts.info_log) {
+    s = CreateLoggerFromOptions(dbname, tmp_opts, &tmp_opts.info_log);
+    if (!s.ok()) {
+      tmp_opts.info_log = nullptr;
+      return s;
+    }
+  }
+
+  handles->clear();
+  DBImplFollower* impl =
+      new DBImplFollower(tmp_opts, std::move(new_env), dbname, src_path);
+  impl->versions_.reset(new ReactiveVersionSet(
+      dbname, &impl->immutable_db_options_, impl->file_options_,
+      impl->table_cache_.get(), impl->write_buffer_manager_,
+      &impl->write_controller_, impl->io_tracer_));
+  impl->column_family_memtables_.reset(
+      new ColumnFamilyMemTablesImpl(impl->versions_->GetColumnFamilySet()));
+  impl->wal_in_db_path_ = impl->immutable_db_options_.IsWalDirSameAsDBPath();
+
+  impl->mutex_.Lock();
+  s = impl->Recover(column_families, true, false, false);
+  if (s.ok()) {
+    for (const auto& cf : column_families) {
+      auto cfd =
+          impl->versions_->GetColumnFamilySet()->GetColumnFamily(cf.name);
+      if (nullptr == cfd) {
+        s = Status::InvalidArgument("Column family not found", cf.name);
+        break;
+      }
+      handles->push_back(new ColumnFamilyHandleImpl(cfd, impl, &impl->mutex_));
+    }
```

Comment:
Its not a requirement at this point. I'm not sure how difficult it would be. Maybe its trivial (we do allow that if the DB is opened in read-only mode), but I haven't thought about it too much.